passes/pmgen/split_shiftx.pmg

```diff
@@ -0,0 +1,59 @@
+state <SigSpec> shiftxB
+
+match shiftx
+	select shiftx->type == $shiftx
+	select param(shiftx, \Y_WIDTH).as_int() > 1
+endmatch
+
+match macc
+	select macc->type == $macc
+	select param(macc, \B_WIDTH).as_int() == 0
+	optional
+endmatch
+
+code shiftxB
+	shiftxB = port(shiftx, \B);
+
+	if (macc) {
+		const int b_width = param(shiftx, \B_WIDTH).as_int();
+		if (param(shiftx, \B_SIGNED) != 0 && shiftxB[b_width-1] == RTLIL::S0)
+			shiftxB = shiftxB.extract(0, b_width-1);
+
+		if (port(macc, \Y) != shiftxB) {
+			blacklist(shiftx);
+			reject;
+		}
+
+		Const config = param(macc, \CONFIG);
+		const int config_width = param(macc, \CONFIG_WIDTH).as_int();
+		const int num_bits = config.extract(0, 4).as_int();
+		const int num_ports = (config_width - 4) / (2 + 2*num_bits);
+		if (num_ports != 1) {
+			blacklist(shiftx);
+			reject;
```

Comment:
`num_ports != 1` doesn't depend on shiftx or any other state, just on the macc cell. So this should be a `select` statement inside of `match macc` so it is only done once when the pattern matcher initializes.

I also don't understand why you are blacklisting shiftx here.